api/client.cpp

```diff
@@ -0,0 +1,99 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include <api/client.hpp>
+
+#include <common/error.hpp>
+#include <common/expected.hpp>
+#include <common/http.hpp>
+#include <common/log.hpp>
+
+namespace mender {
+namespace api {
+
+namespace error = mender::common::error;
+namespace expected = mender::common::expected;
+namespace http = mender::http;
+namespace log = mender::common::log;
+
+error::Error Client::AsyncCall(
+	http::OutgoingRequestPtr req,
+	http::ResponseHandler header_handler,
+	http::ResponseHandler body_handler) {
+	auto authenticating_handler =
+		[this, req, header_handler, body_handler](auth::ExpectedToken ex_tok) {
+			if (!ex_tok) {
+				log::Error("Failed to obtain authentication credentials");
+				event_loop_.Post([header_handler, ex_tok]() {
+					error::Error err = ex_tok.error();
+					header_handler(expected::unexpected(err));
+				});
+				event_loop_.Post([body_handler, ex_tok]() {
+					error::Error err = ex_tok.error();
+					body_handler(expected::unexpected(err));
+				});
```

Comment:
So if you do `http::Client.AsyncCall()` and you get `401` or `404` or whatever error from the server, the body handler is not called? I thought it was. This should definitely behave the same way.

Reply:
Yes, in that case both should be called. Maybe I'm a bit confused about this code, but you are checking for `if (ex_tok)`, which indicates an error other than 401 and 404 (which do not produce errors, but just "successful" 4xx status codes).

I guess my point is that `header_handler` and `body_handler` should not both be called with an error which is non-null. Whether this actually resulted from a real error, or from special handling of a 401 code, is not that important.

Reply:
`ex_tok` being `false` means authentication, i.e. getting a new token, failed. For whatever reason, but I don't think this code should care about the exact reason. 

And then this is about the generic case -- should `http::Client.AsyncCall()` (including the overriding versions) **always** call both handlers? I guess the current implementation in the base class does not do that in case of some errors (I/O, DNS resolution,...)? To me it makes the API nicer/easier to use if you have a guarantee that both of the handlers will be called. Otherwise there must be a logic in the header handler making sure it does some work from the body handler in cases in which the body handler won't be called. Even just a trivial example -- we do `loop.Stop()` in unit tests in body handlers, if they are sometimes not called, we need to cover those cases in the respective header handlers.

Reply:
Hmm, maybe. Then I suppose you would never really handle errors in the header handler, just bail out early, and then do all handling in the body handler? Could be..

Or if we look it not as "header and body handler", but as "header and finished handler", we could say that only the "finished handler" is called on error, which would eliminate errors from the header handler entirely.

But either of these is a big change to make now though. It changes the logic of the whole API, and I'm a bit reluctant to do it now that we are already using it in many places, and in a lot of tests.

> `ex_tok` being `false` means authentication, i.e. getting a new token, failed. For whatever reason, but I don't think this code should care about the exact reason.

I agree, and it doesn't need to. It just needs to make sure the downstream handlers are not called with more than one error.

Reply:
> But either of these is a big change to make now though. It changes the logic of the whole API, and I'm a bit reluctant to do it now that we are already using it in many places, and in a lot of tests.

Fair enough.